packages/project-account-switcher/src/behavior/ProjectAccountSwitcherBehavior.js

```diff
@@ -0,0 +1,115 @@
+import { Component } from "react";
+import PropTypes from "prop-types";
+
+export default class ProjectAccountSwitcherBehavior extends Component {
+  static propTypes = {
+    /** List of Accounts */
+    accounts: PropTypes.arrayOf(
+      PropTypes.shape({
+        id: PropTypes.string,
+        image: PropTypes.string,
+        label: PropTypes.string
+      })
+    ),
+    /** Currently selected Account */
+    activeAccountId: PropTypes.string,
+    /** Currently selected Project */
+    activeProjectId: PropTypes.string,
+    /** Render prop */
+    children: PropTypes.func,
+    /** Initially sets the selected Account, but allows user action to change it */
+    defaultActiveAccountId: PropTypes.string,
+    /** Initially sets the selected Project, but allows user action to change it */
+    defaultActiveProjectId: PropTypes.string,
+    /** Called when a user clicks on the element */
+    onClick: PropTypes.func,
+    /** Called when a user clicks on the target element */
+    onTargetClick: PropTypes.func,
+    /** Opens the Flyout */
+    open: PropTypes.bool,
+    /** List of Projects */
+    projects: PropTypes.arrayOf(
+      PropTypes.shape({
+        id: PropTypes.string,
+        image: PropTypes.string,
+        label: PropTypes.string
+      })
+    )
+  };
+
+  static defaultProps = {
+    defaultActiveAccountId: "1",
+    defaultActiveProjectId: "1",
+    open: false
+  };
+
+  state = {
+    activeAccountId: this.props.defaultActiveAccountId,
+    activeProjectId: this.props.defaultActiveProjectId,
+    open: this.props.open
+  };
+
+  isControlled() {
+    return (
+      this.props.activeAccountId !== undefined &&
+      this.props.activeProjectId !== undefined
+    );
+  }
+
+  handleTargetClick = event => {
+    const { onTargetClick } = this.props;
+
+    if (onTargetClick) {
+      onTargetClick(event);
+    }
+
+    this.setState({
+      open: !this.state.open
+    });
+  };
+
+  handleAccountClick = event => {
+    const { onClick } = this.props;
+
+    if (onClick) {
+      onClick(event);
+    }
+
+    if (!this.isControlled()) {
+      const selectedAccount = event.target.dataset.accountId;
+      this.setState({
+        activeAccountId: this.props.accounts.filter(
```

Comment:
Notice this filter is repeated.  Here, https://github.com/Autodesk/hig/pull/1102/files#diff-76b409a9682299bea2542680188dfc48R66 and https://github.com/Autodesk/hig/pull/1102/files#diff-76b409a9682299bea2542680188dfc48R78

I wonder if it would make sense to have `ProjectAccountSwitcher` pass the account/project object to the behavior and presenter classes?  Rather than the id

Reply:
I'm with you; I found this repetition bothersome as well but wasn't sure what refactoring for some kind of Account/Project object look like. My thought was that it would be another component, and it felt a little premature to make that abstraction since I don't see this account/project pattern elsewhere. 

I think input from @nfiniteset @morrisallison and @wmui51 would be helpful here for deeper HIG context.

And/or please let me know if I'm misunderstanding what this refactor might look like.

Reply:
I'm not completely sure -- I had the idea based off of the way the vanilla component was structured: https://github.com/Autodesk/hig/tree/development/packages/vanilla/src/components/global-nav/top-nav/project-account-switcher

I think that the example you wrote out above makes sense though!

Reply:
@arirusso I'm working through the change you proposed, but am unsure how the behavioral component will be able to connect a click on the markup that represents each `account` or `project` without passing through the `id` in the HTML element's dataset and using it to look up an `account`/`project` in `accounts`/`projects`.

Right now I'm trying to get the meaningful information within the event listeners defined in the behavioral component _without_ referring to `accounts`/`projects` and I'm having trouble. 

I was hoping to be able to refer to the `account`/`project` object by just calling `this.event.target` but `this` is undefined in the context of the click listener. `event.target` returns markup like:

```
<span class="project-account-switcher__item__label" data-account-id="1">Account 1</span>
```

Reply:
@halfghaninne 

Happy to work with you on this later on if it's helpful.  I'll try and type it out but it's a mouthful and may be easier IRL: 

I think there's a few issues to consider here

One - It's not entirely clear what the purpose of the filter being used in the event here is.  Is it to ensure that the account id is valid?  Couldn't in this case, this filter code be written like this?

```
this.setState({
        activeAccountId: selectedAccount
});
```

Next, even before implementing something like the API sketch I did above, the filter is repeated a handful of times - that could at least be extracted into a helper method to DRY things up.  Something like:

```
this.setState({
        activeAccountId: getActiveAccountId(this.props)
});

This stage could continue to use account/project ids.  

Then finally, once things are consolidated and cleaned up, you could move to using accounts/projects instead of passing around ids as above. The events would probably still use `data-account-id` using `account.id` or whatever. But the API of the React components would change to use objects

Reply:
Thanks! Between your response and getting a better understanding of my options for passing info to the event listener, I addressed this in: c573b26